ns for fx: rekey results by model node names

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60419 ns for fx: initial support for int8 shadows fp32
* #60559 ns for fx: fix shadow logger errors for resnet18 and mobilenet_v2
* #60311 ns for fx: expose function to add comparisons between logged values
* **#60305 ns for fx: rekey results by model node names**

Summary:

Adjusts the NS for FX weight and activation extraction APIs
to require a model name, and rekeys the results of these APIs
to use the node names of the specified model as layer keys.

This is useful because the "before" layer names are not very meaningful
to users, they were chosen just to guarantee consistency between
different models.  The new layer names correspond to the real locations
in one of the models.

For example, before

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger)

// results
{'base_op_1_0': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

and after

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger, 'model_b_name')

// results
// note: instead of `base_op_1_0`, the layer is named `linear1`
{'linear1': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

Note: we cannot use these names while collecting data because
node names are not guaranteed to be consistent across graphs.
This is why we only rekey as the very last step.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_layer_names
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29243045](https://our.internmc.facebook.com/intern/diff/D29243045)

vkuzo committed:
ns for fx: rekey results by model node names

Summary:

Adjusts the NS for FX weight and activation extraction APIs
to require a model name, and rekeys the results of these APIs
to use the node names of the specified model as layer keys.

For example, before

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger)

// results
{'base_op_1_0': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

and after

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger, 'model_b_name')

// results
// note: instead of `base_op_1_0`, the layer is named `linear1`
{'linear1': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

Note: we cannot use these names while collecting data because
node names are not guaranteed to be consistent across graphs.
This is why we only rekey as the very last step.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_layer_names
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29243045).

hx89 commented on torch/quantization/ns/utils.py:
```diff
@@ -227,3 +227,41 @@ def get_target_type_str(node: Node, gm: GraphModule) -> str:
+          [{'ref_node_name': 'linear1', ...}]}}}
+
+    Note: we cannot use these node names directly because they are not
+    guaranteed to be consistent across models. This is why we extract
```
Just wonder in which cases the node names are not consistent across models? Could you add an example if possible?

vkuzo replied:
As far as I know there are no guarantees provided by FX on what node names are generated when tracing through the model.  They are the same today, but that could break at any time in the future, and it's not something we can depend on.

hx89 commented on torch/quantization/ns/utils.py:
```diff
@@ -227,3 +227,41 @@ def get_target_type_str(node: Node, gm: GraphModule) -> str:
+    into
+
+        {'linear1': {'node_output': {'model_a':
+          [{'ref_node_name': 'linear1', ...}]}}}
```
Is it possible to keep the old layer name in the result? I can't think of specific use case now, but feel it might be useful during debugging.

hx89 replied:
Since we already use ref_node_name as layer name, does it make sense to remove "ref_node_name" field from the result to reduce duplication?

hx89 replied:
> Is it possible to keep the old layer name in the result? I can't think of specific use case now, but feel it might be useful during debugging.

Sorry seems we have "prev_node_name" in the result, please ignore this comment.

vkuzo replied:
> Is it possible to keep the old layer name in the result? I can't think of specific use case now, but feel it might be useful during debugging.

yes, the old layer name is unchanged in the result

> Since we already use ref_node_name as layer name, does it make sense to remove "ref_node_name" field from the result to reduce duplication?

hmm, so `ref_node_name` of one of the models is used as layer_name.  Might be good to keep it in the result as well, just to not change the keys of what is stored in the result.

vkuzo committed:
Update on "ns for fx: rekey results by model node names"


Summary:

Adjusts the NS for FX weight and activation extraction APIs
to require a model name, and rekeys the results of these APIs
to use the node names of the specified model as layer keys.

This is useful because the "before" layer names are not very meaningful
to users, they were chosen just to guarantee consistency between
different models.  The new layer names correspond to the real locations
in one of the models.

For example, before

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger)

// results
{'base_op_1_0': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

and after

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger, 'model_b_name')

// results
// note: instead of `base_op_1_0`, the layer is named `linear1`
{'linear1': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

Note: we cannot use these names while collecting data because
node names are not guaranteed to be consistent across graphs.
This is why we only rekey as the very last step.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_layer_names
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29243045](https://our.internmc.facebook.com/intern/diff/D29243045)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29243045).

vkuzo committed:
Update on "ns for fx: rekey results by model node names"


Summary:

Adjusts the NS for FX weight and activation extraction APIs
to require a model name, and rekeys the results of these APIs
to use the node names of the specified model as layer keys.

This is useful because the "before" layer names are not very meaningful
to users, they were chosen just to guarantee consistency between
different models.  The new layer names correspond to the real locations
in one of the models.

For example, before

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger)

// results
{'base_op_1_0': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

and after

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger, 'model_b_name')

// results
// note: instead of `base_op_1_0`, the layer is named `linear1`
{'linear1': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

Note: we cannot use these names while collecting data because
node names are not guaranteed to be consistent across graphs.
This is why we only rekey as the very last step.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_layer_names
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29243045](https://our.internmc.facebook.com/intern/diff/D29243045)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29243045).

vkuzo committed:
Update on "ns for fx: rekey results by model node names"


Summary:

Adjusts the NS for FX weight and activation extraction APIs
to require a model name, and rekeys the results of these APIs
to use the node names of the specified model as layer keys.

This is useful because the "before" layer names are not very meaningful
to users, they were chosen just to guarantee consistency between
different models.  The new layer names correspond to the real locations
in one of the models.

For example, before

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger)

// results
{'base_op_1_0': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

and after

```
// API call
results = ns.extract_logger_info(
  model_a, model_b, ns.OutputLogger, 'model_b_name')

// results
// note: instead of `base_op_1_0`, the layer is named `linear1`
{'linear1': {'node_output':
  {'model_a': [{'ref_node_name': 'linear1', ...}]}}}
```

Note: we cannot use these names while collecting data because
node names are not guaranteed to be consistent across graphs.
This is why we only rekey as the very last step.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_layer_names
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29243045](https://our.internmc.facebook.com/intern/diff/D29243045)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29243045).
